### PR TITLE
Use `create` method for object construction only

### PR DESCRIPTION
**Pull request by hsorby**
This work forces the use of the `create` method for constructing new objects for all classes in the library.

Addresses #458.

### CODE REVIEW

**Comment by hsorby** [comment]
Ready for review.

File: src/api/libcellml/component.h [hunk 1]
```diff
@@ -43,11 +43,26 @@ class LIBCELLML_EXPORT Component: public ComponentEntity, public ImportedEntity
...
+     * @brief Create a @c Component object.
+     *
+     * Factory method to create a @c Component.  Create a
+     * blank component with::
+     *
+     *   ComponentPtr component = libcellml::Component::create();
+     *
+     * or a named component with name "Component" with::
+     *
+     *   ComponentPtr component = libcellml::Component::create(std::string("Component"));
```
Comment (agarny) [inline src/api/libcellml/component.h]: Might want to simplify this a bit:

```
ComponentPtr component = libcellml::Component::create("Component");
```

File: src/api/libcellml/model.h [hunk 2]
```diff
@@ -44,11 +44,26 @@ class LIBCELLML_EXPORT Model: public ComponentEntity
...
+     * @brief Create a @c Model object.
+     *
+     * Factory method to create a @c Model.  Create a
+     * blank model with::
+     *
+     *   ModelPtr model = libcellml::Model::create();
+     *
+     * or a named model with name "Model" with::
+     *
+     *   ModelPtr model = libcellml::Model::create(std::string("Model"));
```
Comment (agarny) [inline src/api/libcellml/model.h]: Simplify with:

```
ModelPtr model = libcellml::Model::create("Model");
```

File: src/api/libcellml/reset.h [hunk 3]
```diff
@@ -37,11 +37,26 @@ class LIBCELLML_EXPORT Reset: public OrderedEntity
...
-    }
+    /**
+     * @brief Create a @c Reset object.
+     *
+     * Factory method to create a @c Reset.  Create a
+     * blank reset with::
+     *
+     *   ResetPtr reset = libcellml::Reset::create();
+     *
+     * or an ordered reset with order "10" with::
```
Comment (agarny) [inline src/api/libcellml/reset.h]: Drop the double quotes, it's not a string.

File: src/api/libcellml/variable.h [hunk 4]
```diff
@@ -37,11 +37,26 @@ class LIBCELLML_EXPORT Variable: public NamedEntity
...
+     * @brief Create a @c Variable object.
+     *
+     * Factory method to create a @c Variable.  Create a
+     * blank variable with::
+     *
+     *   VariablePtr variable = libcellml::Variable::create();
+     *
+     * or a named variable with name "Variable" with::
+     *
+     *   VariablePtr variable = libcellml::Variable::create("Variable");
```
Comment (agarny) [inline src/api/libcellml/variable.h:50]: Simplify with:

```
VariablePtr variable = libcellml::Variable::create("Variable");
```

**Review by agarny — CHANGES_REQUESTED** [review]
Looks good, pending some (very) minor issues.

**Review by agarny — APPROVED** [review]

**Comment by nickerso** [comment]
Its not clear to me why different objects have different `create` methods. It maybe makes sense that classes derived from `NamedEntity` have a create method with a `name` argument, but why a `Reset` create with an `order` attribute? Or perhaps why no `Variable` create method with name and units? (or did I miss it?)

**Comment by hsorby** [comment]
> Its not clear to me why different objects have different `create` methods. It maybe makes sense that classes derived from `NamedEntity` have a create method with a `name` argument, but why a `Reset` create with an `order` attribute? Or perhaps why no `Variable` create method with name and units? (or did I miss it?)

I have added some convenience methods to see how they would work.  A `Reset` must have an order set so it seemed useful to have a method that allowed the creation of a `Reset` with an order.

I haven't done all possible convenience methods just the ones that I thought about as I modified the `create` method.  'Missing' convenience methods can be added at any time for those that have been missed.

**Review by nickerso — APPROVED** [review]

**Comment by kerimoyle** [comment]
I'd have thought that having a reset with the test_variable and the reset_variable made more sense as a constructor than the order alone ... but as you say, there are a whole load of combinations ...

**Comment by hsorby** [comment]
These are all good ideas we should probably put them into an issue on convenience constructors :)

**Pull request merged**